### DISCUSSION TITLE
Out of the box support for Web Socket

### DISCUSSION THREAD

**Discussion by cchirag** [vercel/next.js#65469] [Ideas]
### Goals

1. To support Web Sockets out of the box.
2. Like Route Handlers, create a file (something like event.js, message.js or socket.js) like route.js and page.js. This is then injected into the socket's on(message) handler.
3. Vercel deployment support. Currently, to create a websocket endpoint, we need to create an external server which won't get deployed on Vercel and breaks other optimisations.

### Non-Goals

1. It does not aim to address custom implementation of websocket. Meaning, if we choose to implement the feature with Socket.io, then the user does not have control over changing this implementation.

### Background

I feel this is a necessary feature because:
- There are many blogs about creating a websocket endpoint. But these implementations are crude and not standardised.
- Ease of implementing realtime data transfer needs like a chat application.

With the current method of implementation, one needs to create a separate server. 
These implementations have few drawbacks like:
- Bad DX
- Loss of optimisation that Next JS provides.
- Unable to deploy application on Vercel.

### Proposal

This feature can be implemented by:
- Adding WS feature into the native node JS server and exposing it in Dev and Production builds.
- Compiling and bundling the WS message code into the native Node JS server.

I'm not really sure if anything else needs to be changed on Vercel but I would love to work on this and see where it takes us

**Comment by zhoupb01** [comment]
Hello,

How do you use Web Sockets to implement a real-time chat feature?

Do you use an Express service and then use `useEffect` in a client component to connect to the Express service?

**Comment by AlessandroFare** [comment]
There isn't really native long-lived WS on Vercel serverless like on a Node server or on Cloudflare Workers + Durable Objects.
People usually: separate WS host + JWT from Route Handlers, or Pusher/Ably, or a CF Worker room layer with the SDK in client components.

Chat with history, or one-way alerts?
